depthcharge/data/parsers.py

```diff
@@ -531,14 +533,150 @@ def parse_spectrum(self, spectrum: dict) -> MassSpectrum:
+            )
+        if custom_fields is not None:
+            warnings.warn(
+                "custom_fields is currently not supported for "
+                "TdfParser. Continuing with default value `None`.",
+                UserWarning,
+                stacklevel=2,
+            )
+            custom_fields = None
+        super().__init__(
+            peak_file,
+            ms_level=ms_level,
+            preprocessing_fn=preprocessing_fn,
+            valid_charge=valid_charge,
+            custom_fields=custom_fields,
+            progress=progress,
+            id_type="index",
+        )
+        self._counter = -1
+
+    def sniff(self) -> None:
+        """Quickly test a file for the correct type.
+
+        Raises
+        ------
+        IOError
+            Raised if the file is not the expected format.
+
+        """
+        if (
+            not self.peak_file.exists()
+            or self.peak_file.suffix.lower() != ".d"
+        ):
+            raise OSError("Not a TDF file.")
+        try:
+            timsrust_pyo3.SpectrumReader(fspath(self.peak_file))
+        except OSError:
+            raise OSError("Not a TDF file.")
+
+    def _spectrum_to_dict(self, spectrum) -> dict:
+        """Convert a Spectrum into a plain dict using the known schema."""
+        p = spectrum.precursor
+        return {
+            "index": spectrum.index,
+            "mz_values": list(spectrum.mz_values),
+            "intensities": list(spectrum.intensities),
+            "precursor": {
+                "mz": p.mz,
+                "rt": p.rt,
+                "im": p.im,
+                "charge": p.charge,
+                "intensity": p.intensity,
+            },
+            "collision_energy": spectrum.collision_energy,
+            "isolation_mz": spectrum.isolation_mz,
+            "isolation_width": spectrum.isolation_width,
+        }
+
+    @contextmanager
+    def open(self):
+        """Open the TDF file for reading."""
+        reader = timsrust_pyo3.SpectrumReader(fspath(self.peak_file))
+
+        def _iter():
+            n = len(reader)
+            for i in range(n):
+                try:
+                    spec = reader.get(i)
+                    yield self._spectrum_to_dict(spec)
+                except OSError:
+                    continue  # skip broken spectra
+
+        yield _iter()
+
+    def parse_spectrum(self, spectrum: dict) -> MassSpectrum:
+        """Parse a single spectrum.
+
+        Parameters
+        ----------
+        spectrum : dict
+            The dictionary defining the spectrum in TDF format.
+
+        """
+        self._counter += 1
+
+        precursor_mz = float(spectrum["precursor"]["mz"])
+        precursor_charge = float(spectrum["precursor"]["charge"])
+
+        if self.valid_charge is None or precursor_charge in self.valid_charge:
+            return MassSpectrum(
+                filename=str(self.peak_file),
+                scan_id=spectrum["index"],
```

Comment:
I am not 100% sure what the index is meant to be ... so I cannot assure this is correct. In essence ... the notion of a scan_id is very fluid for tof data ... 

Since every spectrum is in essence the aggregate of multiple scans within a frame ... so the "real" identifier would be something like Frame=32, scan_index=102-150, and this scan index is per-frame, not per-file ... AND that gets indexed in the SpectrumReader as just sequential ...

(I would suggest adding a comment making clear what the id means)